grails-app/controllers/org/pih/warehouse/inventory/InventoryController.groovy

```diff
@@ -575,49 +575,70 @@ class InventoryController {
 
 
     def listExpiredStock() {
-        def location = Location.get(session.warehouse.id)
-        def categorySelected = (params.category) ? Category.get(params.category) : null
-        def inventoryItems = dashboardService.getExpiredStock(categorySelected, location)
-        def categories = inventoryItems?.collect { it.product.category }?.unique()
-        def quantityMap = productAvailabilityService.getQuantityOnHandByInventoryItem(location)
-        def expiredStockMap = [:]
-        inventoryItems.each { inventoryItem ->
-            expiredStockMap[inventoryItem] = quantityMap[inventoryItem]
+        Location location = Location.get(session.warehouse.id)
+        Category categorySelected = Category.get(params.category)
+        Boolean withBinLocation = params.boolean("withBinLocation")
+
+        List<InventoryItem> inventoryItems = dashboardService.getExpiredStock(categorySelected, location)
+        List<Category> categories = inventoryItems?.collect { it.product.category }?.unique()
+
+        List<Map> data
+        if (withBinLocation) {
+            data = productAvailabilityService.getQuantityOnHandByBinLocation(location)
+        } else {
+            data = productAvailabilityService.getQuantityOnHandByInventoryItem(location)
+                    .collect{ key, val -> [ inventoryItem: key, quantity: val ] }
         }
+        data = data.findAll { inventoryItems.contains(it.inventoryItem) }
```

Comment:
Can't we do this filtering in the PA query? 
I think we could just pass the `inventoryItems` (or `inventoryItems.id`) to `getQuantityByBinLocation` (or the second method as well) and just add the `IN :inventoryItems` clause?

This could additionally improve the performance.

And a suggestion (regarding what version you prefer):
```groovy
List<Map> data = withBinLocation 
   ? productAvailabilityService.getQuantityOnHandByBinLocation(location) 
   : productAvailabilityService.getQuantityOnHandByInventoryItem(location)
         .collect{ key, val -> [ inventoryItem: key, quantity: val ] }
```

Reply:
Sure I can do it.
I just didn't want to modify the old service methods for that but it probably is worth changing it.
I was following the old logic for getting the expired or expiring inventoryItems and filtering the PA by that but I agree, this should probably be done in the db.

Reply:
One thing to mention is that we should probably have this type of logic in a more central service, like ProductAvailabilityService. The DashboardService was created before our product availability and should probably be phased out or refactored (i.e. it should delegate to product availability service and other reporting services).